"""
–°–µ—Ä—å—ë–∑–Ω—ã–µ unit —Ç–µ—Å—Ç—ã –¥–ª—è –∫–æ–Ω–≤–µ—Ä—Ç–µ—Ä–æ–≤ DTO ‚Üî protobuf.
–ü—Ä–æ–≤–µ—Ä—è—é—Ç –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ—Å—Ç—å –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏, –æ–±—Ä–∞–±–æ—Ç–∫—É –æ—à–∏–±–æ–∫, —Å–æ–≤–º–µ—Å—Ç–∏–º–æ—Å—Ç—å.
"""

import pytest
import uuid
from unittest.mock import patch

from ..conv import (
    dto_to_proto,
    proto_to_dto,
    transition_dto_to_proto,
    transition_proto_to_dto,
    dto_to_json_dict,
    dto_to_protobuf_json,
    ConversionError,
    create_proto_snapshot,
    parse_proto_snapshot,
    FSM_STATE_DTO_TO_PROTO,
    FSM_STATE_PROTO_TO_DTO,
)
from ..types import (
    FsmSnapshotDTO,
    TransitionDTO,
    FsmState,
    TransitionStatus,
    initial_snapshot,
    create_transition,
)

# –ò–º–ø–æ—Ä—Ç protobuf –¥–ª—è —Ç–µ—Å—Ç–æ–≤
from generated.fsm_state_pb2 import (
    FsmStateSnapshot,
    StateTransition,
    FSMStateEnum,
    FSMTransitionStatus,
)


class TestEnumMappings:
    """–¢–µ—Å—Ç—ã –º–∞–ø–ø–∏–Ω–≥–∞ enum'–æ–≤ –º–µ–∂–¥—É DTO –∏ protobuf"""

    def test_fsm_state_dto_to_proto_mapping(self):
        """–¢–µ—Å—Ç –º–∞–ø–ø–∏–Ω–≥–∞ FsmState -> FSMStateEnum"""
        assert (
            FSM_STATE_DTO_TO_PROTO[FsmState.UNSPECIFIED]
            == FSMStateEnum.FSM_STATE_UNSPECIFIED
        )
        assert FSM_STATE_DTO_TO_PROTO[FsmState.BOOTING] == FSMStateEnum.BOOTING
        assert FSM_STATE_DTO_TO_PROTO[FsmState.IDLE] == FSMStateEnum.IDLE
        assert FSM_STATE_DTO_TO_PROTO[FsmState.ACTIVE] == FSMStateEnum.ACTIVE
        assert FSM_STATE_DTO_TO_PROTO[FsmState.ERROR_STATE] == FSMStateEnum.ERROR_STATE
        assert FSM_STATE_DTO_TO_PROTO[FsmState.SHUTDOWN] == FSMStateEnum.SHUTDOWN

    def test_fsm_state_proto_to_dto_mapping(self):
        """–¢–µ—Å—Ç –º–∞–ø–ø–∏–Ω–≥–∞ FSMStateEnum -> FsmState"""
        assert (
            FSM_STATE_PROTO_TO_DTO[FSMStateEnum.FSM_STATE_UNSPECIFIED]
            == FsmState.UNSPECIFIED
        )
        assert FSM_STATE_PROTO_TO_DTO[FSMStateEnum.BOOTING] == FsmState.BOOTING
        assert FSM_STATE_PROTO_TO_DTO[FSMStateEnum.IDLE] == FsmState.IDLE
        assert FSM_STATE_PROTO_TO_DTO[FSMStateEnum.ACTIVE] == FsmState.ACTIVE
        assert FSM_STATE_PROTO_TO_DTO[FSMStateEnum.ERROR_STATE] == FsmState.ERROR_STATE
        assert FSM_STATE_PROTO_TO_DTO[FSMStateEnum.SHUTDOWN] == FsmState.SHUTDOWN

    def test_bidirectional_enum_mapping(self):
        """–¢–µ—Å—Ç —á—Ç–æ –º–∞–ø–ø–∏–Ω–≥ —Ä–∞–±–æ—Ç–∞–µ—Ç –≤ –æ–±–µ —Å—Ç–æ—Ä–æ–Ω—ã"""
        for dto_state, proto_state in FSM_STATE_DTO_TO_PROTO.items():
            assert FSM_STATE_PROTO_TO_DTO[proto_state] == dto_state


class TestTransitionConversion:
    """–¢–µ—Å—Ç—ã –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ –ø–µ—Ä–µ—Ö–æ–¥–æ–≤ TransitionDTO ‚Üî StateTransition"""

    def test_transition_dto_to_proto(self):
        """–¢–µ—Å—Ç –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ TransitionDTO -> StateTransition"""
        dto = TransitionDTO(
            from_state=FsmState.IDLE,
            to_state=FsmState.ACTIVE,
            trigger_event="PROPOSALS_RECEIVED",
            status=TransitionStatus.SUCCESS,
            error_message="",
            ts_wall=1234567890.5,
        )

        proto = transition_dto_to_proto(dto)

        assert proto.from_state == FSMStateEnum.IDLE
        assert proto.to_state == FSMStateEnum.ACTIVE
        assert proto.trigger_event == "PROPOSALS_RECEIVED"
        assert proto.status == FSMTransitionStatus.SUCCESS
        assert proto.error_message == ""
        assert proto.timestamp.seconds == 1234567890
        assert proto.timestamp.nanos == 500000000  # 0.5 —Å–µ–∫ –≤ –Ω–∞–Ω–æ—Å–µ–∫—É–Ω–¥–∞—Ö

    def test_transition_proto_to_dto(self):
        """–¢–µ—Å—Ç –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ StateTransition -> TransitionDTO"""
        proto = StateTransition()
        proto.from_state = FSMStateEnum.ACTIVE
        proto.to_state = FSMStateEnum.IDLE
        proto.trigger_event = "NO_PROPOSALS"
        proto.status = FSMTransitionStatus.SUCCESS
        proto.error_message = ""
        proto.timestamp.FromSeconds(1234567890)
        proto.timestamp.nanos = 250000000  # 0.25 —Å–µ–∫

        dto = transition_proto_to_dto(proto)

        assert dto.from_state == FsmState.ACTIVE
        assert dto.to_state == FsmState.IDLE
        assert dto.trigger_event == "NO_PROPOSALS"
        assert dto.status == TransitionStatus.SUCCESS
        assert dto.error_message == ""
        assert abs(dto.ts_wall - 1234567890.25) < 0.001  # –ø—Ä–∏–º–µ—Ä–Ω–æ —Ä–∞–≤–Ω–æ

    def test_transition_with_error(self):
        """–¢–µ—Å—Ç –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ –ø–µ—Ä–µ—Ö–æ–¥–∞ —Å –æ—à–∏–±–∫–æ–π"""
        dto = TransitionDTO(
            from_state=FsmState.IDLE,
            to_state=FsmState.ERROR_STATE,
            trigger_event="BIOS_ERROR",
            status=TransitionStatus.FAILED,
            error_message="BIOS system check failed",
        )

        proto = transition_dto_to_proto(dto)
        back_dto = transition_proto_to_dto(proto)

        assert back_dto.from_state == dto.from_state
        assert back_dto.to_state == dto.to_state
        assert back_dto.trigger_event == dto.trigger_event
        assert back_dto.status == dto.status
        assert back_dto.error_message == dto.error_message

    def test_transition_roundtrip(self):
        """–¢–µ—Å—Ç roundtrip –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ –ø–µ—Ä–µ—Ö–æ–¥–∞"""
        original = create_transition(
            FsmState.BOOTING, FsmState.IDLE, "BOOT_COMPLETE", TransitionStatus.SUCCESS
        )

        proto = transition_dto_to_proto(original)
        converted_back = transition_proto_to_dto(proto)

        # –ü—Ä–æ–≤–µ—Ä—è–µ–º –æ—Å–Ω–æ–≤–Ω—ã–µ –ø–æ–ª—è (–≤—Ä–µ–º–µ–Ω–Ω—ã–µ –º–µ—Ç–∫–∏ –º–æ–≥—É—Ç –æ—Ç–ª–∏—á–∞—Ç—å—Å—è)
        assert converted_back.from_state == original.from_state
        assert converted_back.to_state == original.to_state
        assert converted_back.trigger_event == original.trigger_event
        assert converted_back.status == original.status
        assert converted_back.error_message == original.error_message


class TestSnapshotConversion:
    """–¢–µ—Å—Ç—ã –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ —Å–Ω–∞–ø—à–æ—Ç–æ–≤ FsmSnapshotDTO ‚Üî FsmStateSnapshot"""

    def test_dto_to_proto_basic(self):
        """–¢–µ—Å—Ç –±–∞–∑–æ–≤–æ–π –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ DTO -> protobuf"""
        dto = FsmSnapshotDTO(
            version=42,
            state=FsmState.ACTIVE,
            reason="TEST_REASON",
            snapshot_id=str(uuid.uuid4()),
            fsm_instance_id=str(uuid.uuid4()),
            source_module="test_module",
            attempt_count=5,
        )

        proto = dto_to_proto(dto)

        assert proto.current_state == FSMStateEnum.ACTIVE
        assert proto.source_module == "test_module"
        assert proto.attempt_count == 5
        assert proto.snapshot_id.value == dto.snapshot_id
        assert proto.fsm_instance_id.value == dto.fsm_instance_id

        # DTO-—Å–ø–µ—Ü–∏—Ñ–∏—á–Ω—ã–µ –ø–æ–ª—è –¥–æ–ª–∂–Ω—ã –±—ã—Ç—å –≤ –º–µ—Ç–∞–¥–∞–Ω–Ω—ã—Ö
        assert proto.state_metadata["dto_version"] == "42"
        assert proto.state_metadata["dto_reason"] == "TEST_REASON"

    def test_dto_to_proto_with_history(self):
        """–¢–µ—Å—Ç –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ —Å –∏—Å—Ç–æ—Ä–∏–µ–π –ø–µ—Ä–µ—Ö–æ–¥–æ–≤"""
        transition1 = create_transition(
            FsmState.BOOTING, FsmState.IDLE, "BOOT_COMPLETE"
        )
        transition2 = create_transition(
            FsmState.IDLE, FsmState.ACTIVE, "PROPOSALS_RECEIVED"
        )

        dto = FsmSnapshotDTO(
            version=3, state=FsmState.ACTIVE, history=[transition1, transition2]
        )

        proto = dto_to_proto(dto)

        assert len(proto.history) == 2
        assert proto.history[0].from_state == FSMStateEnum.BOOTING
        assert proto.history[0].to_state == FSMStateEnum.IDLE
        assert proto.history[1].from_state == FSMStateEnum.IDLE
        assert proto.history[1].to_state == FSMStateEnum.ACTIVE

    def test_dto_to_proto_with_metadata(self):
        """–¢–µ—Å—Ç –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ —Å –º–µ—Ç–∞–¥–∞–Ω–Ω—ã–º–∏"""
        context_data = {"sensor_count": "10", "proposal_active": "true"}
        state_metadata = {"debug": "false", "test_mode": "true"}

        dto = FsmSnapshotDTO(
            version=1,
            state=FsmState.IDLE,
            context_data=context_data,
            state_metadata=state_metadata,
        )

        proto = dto_to_proto(dto)

        # –ö–æ–Ω—Ç–µ–∫—Å—Ç–Ω—ã–µ –¥–∞–Ω–Ω—ã–µ
        assert proto.context_data["sensor_count"] == "10"
        assert proto.context_data["proposal_active"] == "true"

        # –ú–µ—Ç–∞–¥–∞–Ω–Ω—ã–µ —Å–æ—Å—Ç–æ—è–Ω–∏—è (–≤–∫–ª—é—á–∞—è DTO-—Å–ø–µ—Ü–∏—Ñ–∏—á–Ω—ã–µ)
        assert proto.state_metadata["debug"] == "false"
        assert proto.state_metadata["test_mode"] == "true"
        assert proto.state_metadata["dto_version"] == "1"

    def test_proto_to_dto_basic(self):
        """–¢–µ—Å—Ç –±–∞–∑–æ–≤–æ–π –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ protobuf -> DTO"""
        proto = FsmStateSnapshot()
        proto.current_state = FSMStateEnum.IDLE
        proto.source_module = "proto_test"
        proto.attempt_count = 7

        # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º UUID
        proto.snapshot_id.value = str(uuid.uuid4())
        proto.fsm_instance_id.value = str(uuid.uuid4())

        # DTO-—Å–ø–µ—Ü–∏—Ñ–∏—á–Ω—ã–µ –º–µ—Ç–∞–¥–∞–Ω–Ω—ã–µ
        proto.state_metadata["dto_version"] = "15"
        proto.state_metadata["dto_reason"] = "PROTO_TEST"
        proto.state_metadata["dto_prev_state"] = "BOOTING"
        proto.state_metadata["dto_ts_mono"] = "1234.567"

        dto = proto_to_dto(proto)

        assert dto.state == FsmState.IDLE
        assert dto.version == 15
        assert dto.reason == "PROTO_TEST"
        assert dto.prev_state == FsmState.BOOTING
        assert dto.source_module == "proto_test"
        assert dto.attempt_count == 7
        assert abs(dto.ts_mono - 1234.567) < 0.001

    def test_roundtrip_conversion(self):
        """–¢–µ—Å—Ç roundtrip –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ —Å–Ω–∞–ø—à–æ—Ç–∞"""
        original = FsmSnapshotDTO(
            version=99,
            state=FsmState.ERROR_STATE,
            prev_state=FsmState.ACTIVE,
            reason="ROUNDTRIP_TEST",
            source_module="roundtrip_module",
            attempt_count=3,
            context_data={"key1": "value1"},
            state_metadata={"key2": "value2"},
        )

        proto = dto_to_proto(original)
        converted_back = proto_to_dto(proto)

        # –û—Å–Ω–æ–≤–Ω—ã–µ –ø–æ–ª—è –¥–æ–ª–∂–Ω—ã —Å–æ–≤–ø–∞–¥–∞—Ç—å
        assert converted_back.version == original.version
        assert converted_back.state == original.state
        assert converted_back.prev_state == original.prev_state
        assert converted_back.reason == original.reason
        assert converted_back.source_module == original.source_module
        assert converted_back.attempt_count == original.attempt_count
        assert converted_back.context_data == original.context_data
        assert converted_back.state_metadata == original.state_metadata

    def test_empty_and_none_handling(self):
        """–¢–µ—Å—Ç –æ–±—Ä–∞–±–æ—Ç–∫–∏ –ø—É—Å—Ç—ã—Ö –∑–Ω–∞—á–µ–Ω–∏–π –∏ None"""
        dto = FsmSnapshotDTO(
            version=0,
            state=FsmState.UNSPECIFIED,
            reason="",
            prev_state=None,
            history=None,  # –¥–æ–ª–∂–Ω–æ —Å—Ç–∞—Ç—å []
            context_data=None,  # –¥–æ–ª–∂–Ω–æ —Å—Ç–∞—Ç—å {}
            state_metadata=None,  # –¥–æ–ª–∂–Ω–æ —Å—Ç–∞—Ç—å {}
        )

        proto = dto_to_proto(dto)
        converted_back = proto_to_dto(proto)

        assert converted_back.reason == ""
        assert converted_back.prev_state is None
        assert converted_back.history == []
        assert converted_back.context_data == {}
        # state_metadata —Å–æ–¥–µ—Ä–∂–∏—Ç DTO-—Å–ø–µ—Ü–∏—Ñ–∏—á–Ω—ã–µ –ø–æ–ª—è, –Ω–æ –ø—É—Å—Ç—ã–µ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–µ


class TestConversionErrors:
    """–¢–µ—Å—Ç—ã –æ–±—Ä–∞–±–æ—Ç–∫–∏ –æ—à–∏–±–æ–∫ –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏"""

    def test_dto_to_proto_with_invalid_enum(self):
        """–¢–µ—Å—Ç –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ —Å –Ω–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω—ã–º enum (–∑–∞—â–∏—Ç–∞ –æ—Ç –±—É–¥—É—â–∏—Ö –∏–∑–º–µ–Ω–µ–Ω–∏–π)"""
        # –°–æ–∑–¥–∞—ë–º DTO —Å "–Ω–æ–≤—ã–º" —Å–æ—Å—Ç–æ—è–Ω–∏–µ–º –∫–æ—Ç–æ—Ä–æ–≥–æ –Ω–µ—Ç –≤ –º–∞–ø–ø–∏–Ω–≥–µ
        dto = FsmSnapshotDTO(
            version=1, state=999, reason="INVALID_ENUM"
        )  # –Ω–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ–µ —Å–æ—Å—Ç–æ—è–Ω–∏–µ

        # –ö–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏—è –¥–æ–ª–∂–Ω–∞ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å fallback
        proto = dto_to_proto(dto)
        assert proto.current_state == FSMStateEnum.FSM_STATE_UNSPECIFIED

    @patch("services.q_core_agent.state.conv.MessageToDict")
    def test_conversion_exception_handling(self, mock_message_to_dict):
        """–¢–µ—Å—Ç –æ–±—Ä–∞–±–æ—Ç–∫–∏ –∏—Å–∫–ª—é—á–µ–Ω–∏–π –≤ –ø—Ä–æ—Ü–µ—Å—Å–µ –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏"""
        mock_message_to_dict.side_effect = Exception("Protobuf error")

        dto = FsmSnapshotDTO(version=1, state=FsmState.IDLE)

        # –î–æ–ª–∂–Ω–æ —É–ø–∞—Å—Ç—å —Å ConversionError
        with pytest.raises(ConversionError):
            dto_to_proto(dto)

    def test_invalid_uuid_handling(self):
        """–¢–µ—Å—Ç –æ–±—Ä–∞–±–æ—Ç–∫–∏ –Ω–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω—ã—Ö UUID"""
        dto = FsmSnapshotDTO(
            version=1,
            state=FsmState.IDLE,
            snapshot_id="invalid-uuid-string",
            fsm_instance_id="",  # –ø—É—Å—Ç–æ–π UUID
        )

        # –ö–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏—è –¥–æ–ª–∂–Ω–∞ –ø—Ä–æ–π—Ç–∏ —Å fallback –Ω–∞ –Ω–æ–≤—ã–µ UUID
        proto = dto_to_proto(dto)

        # UUID –¥–æ–ª–∂–Ω—ã –±—ã—Ç—å –≤–∞–ª–∏–¥–Ω—ã–º–∏
        uuid.UUID(proto.snapshot_id.value)  # –Ω–µ –¥–æ–ª–∂–Ω–æ —É–ø–∞—Å—Ç—å
        uuid.UUID(proto.fsm_instance_id.value)  # –Ω–µ –¥–æ–ª–∂–Ω–æ —É–ø–∞—Å—Ç—å


class TestJSONConversion:
    """–¢–µ—Å—Ç—ã –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ –≤ JSON —Ñ–æ—Ä–º–∞—Ç—ã"""

    def test_dto_to_json_dict(self):
        """–¢–µ—Å—Ç –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ DTO –≤ JSON-—Å–æ–≤–º–µ—Å—Ç–∏–º—ã–π —Å–ª–æ–≤–∞—Ä—å"""
        transition = create_transition(FsmState.BOOTING, FsmState.IDLE, "BOOT")

        dto = FsmSnapshotDTO(
            version=5,
            state=FsmState.ACTIVE,
            prev_state=FsmState.IDLE,
            reason="JSON_TEST",
            history=[transition],
            context_data={"key": "value"},
            state_metadata={"meta": "data"},
        )

        json_dict = dto_to_json_dict(dto)

        assert json_dict["version"] == 5
        assert json_dict["state"] == "ACTIVE"
        assert json_dict["prev_state"] == "IDLE"
        assert json_dict["reason"] == "JSON_TEST"
        assert json_dict["history_count"] == 1
        assert json_dict["context_keys"] == ["key"]
        assert json_dict["metadata_keys"] == ["meta"]
        assert "ts_mono" in json_dict
        assert "ts_wall" in json_dict

    def test_dto_to_protobuf_json(self):
        """–¢–µ—Å—Ç –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ DTO –≤ JSON —á–µ—Ä–µ–∑ protobuf"""
        dto = FsmSnapshotDTO(
            version=3, state=FsmState.ERROR_STATE, reason="PROTOBUF_JSON_TEST"
        )

        json_dict = dto_to_protobuf_json(dto)

        # –î–æ–ª–∂–µ–Ω –±—ã—Ç—å —Å–ª–æ–≤–∞—Ä—å —Å protobuf —Å—Ç—Ä—É–∫—Ç—É—Ä–æ–π
        assert isinstance(json_dict, dict)
        assert (
            "currentState" in json_dict or "current_state" in json_dict
        )  # –º–æ–∂–µ—Ç –≤–∞—Ä—å–∏—Ä–æ–≤–∞—Ç—å—Å—è

    def test_json_formats_consistency(self):
        """–¢–µ—Å—Ç —Å–æ–≥–ª–∞—Å–æ–≤–∞–Ω–Ω–æ—Å—Ç–∏ —Ä–∞–∑–Ω—ã—Ö JSON —Ñ–æ—Ä–º–∞—Ç–æ–≤"""
        dto = initial_snapshot()

        lightweight_json = dto_to_json_dict(dto)
        protobuf_json = dto_to_protobuf_json(dto)

        # –í–µ—Ä—Å–∏–∏ –¥–æ–ª–∂–Ω—ã —Å–æ–≤–ø–∞–¥–∞—Ç—å
        assert lightweight_json["version"] == int(
            protobuf_json.get("stateMetadata", {}).get("dtoVersion", "0")
        )

        # –°–æ—Å—Ç–æ—è–Ω–∏—è –¥–æ–ª–∂–Ω—ã —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤–æ–≤–∞—Ç—å
        assert lightweight_json["state"] == "BOOTING"


class TestHelperFunctions:
    """–¢–µ—Å—Ç—ã –≤—Å–ø–æ–º–æ–≥–∞—Ç–µ–ª—å–Ω—ã—Ö —Ñ—É–Ω–∫—Ü–∏–π"""

    def test_create_proto_snapshot(self):
        """–¢–µ—Å—Ç —Å–æ–∑–¥–∞–Ω–∏—è protobuf —Å–Ω–∞–ø—à–æ—Ç–∞ –∏–∑ –æ—Å–Ω–æ–≤–Ω—ã—Ö –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–≤"""
        proto = create_proto_snapshot(FsmState.ACTIVE, "HELPER_TEST", version=10)

        assert proto.current_state == FSMStateEnum.ACTIVE
        assert proto.state_metadata["dto_reason"] == "HELPER_TEST"
        assert proto.state_metadata["dto_version"] == "10"

    def test_parse_proto_snapshot(self):
        """–¢–µ—Å—Ç –ø–∞—Ä—Å–∏–Ω–≥–∞ protobuf –¥–∞–Ω–Ω—ã—Ö –≤ DTO"""
        # –°–æ–∑–¥–∞—ë–º protobuf —Å–Ω–∞–ø—à–æ—Ç
        original_dto = FsmSnapshotDTO(
            version=7, state=FsmState.SHUTDOWN, reason="PARSE_TEST"
        )
        proto = dto_to_proto(original_dto)
        proto_bytes = proto.SerializeToString()

        # –ü–∞—Ä—Å–∏–º –æ–±—Ä–∞—Ç–Ω–æ
        parsed_dto = parse_proto_snapshot(proto_bytes)

        assert parsed_dto.version == 7
        assert parsed_dto.state == FsmState.SHUTDOWN
        assert parsed_dto.reason == "PARSE_TEST"


class TestTimestampHandling:
    """–¢–µ—Å—Ç—ã –æ–±—Ä–∞–±–æ—Ç–∫–∏ –≤—Ä–µ–º–µ–Ω–Ω—ã—Ö –º–µ—Ç–æ–∫"""

    def test_float_to_timestamp_conversion(self):
        """–¢–µ—Å—Ç –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏ float –≤—Ä–µ–º–µ–Ω–∏ –≤ protobuf Timestamp"""
        from services.q_core_agent.state.conv import (
            _float_to_timestamp,
            _timestamp_to_float,
        )

        # –¢–µ—Å—Ç–∏—Ä—É–µ–º —Ä–∞–∑–ª–∏—á–Ω—ã–µ –≤—Ä–µ–º–µ–Ω–Ω—ã–µ –∑–Ω–∞—á–µ–Ω–∏—è
        test_times = [0.0, 1234567890.5, 1234567890.123456789]

        for original_time in test_times:
            timestamp = _float_to_timestamp(original_time)
            converted_back = _timestamp_to_float(timestamp)

            # –¢–æ—á–Ω–æ—Å—Ç—å –¥–æ –º–∏–ª–ª–∏—Å–µ–∫—É–Ω–¥ –¥–æ–ª–∂–Ω–∞ —Å–æ—Ö—Ä–∞–Ω—è—Ç—å—Å—è
            assert abs(converted_back - original_time) < 0.001

    def test_zero_timestamp_handling(self):
        """–¢–µ—Å—Ç –æ–±—Ä–∞–±–æ—Ç–∫–∏ –Ω—É–ª–µ–≤—ã—Ö –≤—Ä–µ–º–µ–Ω–Ω—ã—Ö –º–µ—Ç–æ–∫"""
        from services.q_core_agent.state.conv import (
            _float_to_timestamp,
            _timestamp_to_float,
        )

        # –ù—É–ª–µ–≤–æ–µ –≤—Ä–µ–º—è
        timestamp = _float_to_timestamp(0.0)
        assert timestamp.seconds == 0
        assert timestamp.nanos == 0

        # –û–±—Ä–∞—Ç–Ω–∞—è –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏—è
        assert _timestamp_to_float(timestamp) == 0.0

        # None timestamp
        assert _timestamp_to_float(None) == 0.0


class TestEdgeCasesAndBoundaries:
    """–¢–µ—Å—Ç—ã –≥—Ä–∞–Ω–∏—á–Ω—ã—Ö —Å–ª—É—á–∞–µ–≤"""

    def test_very_large_version_numbers(self):
        """–¢–µ—Å—Ç –æ—á–µ–Ω—å –±–æ–ª—å—à–∏—Ö –Ω–æ–º–µ—Ä–æ–≤ –≤–µ—Ä—Å–∏–π"""
        large_version = 2**50  # –æ—á–µ–Ω—å –±–æ–ª—å—à–æ–µ —á–∏—Å–ª–æ

        dto = FsmSnapshotDTO(version=large_version, state=FsmState.IDLE)
        proto = dto_to_proto(dto)
        converted_back = proto_to_dto(proto)

        assert converted_back.version == large_version

    def test_unicode_strings_in_conversion(self):
        """–¢–µ—Å—Ç unicode —Å—Ç—Ä–æ–∫ –≤ –∫–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏–∏"""
        unicode_reason = "–¢–µ—Å—Ç –Ω–∞ —Ä—É—Å—Å–∫–æ–º üöÄ ‰∏≠ÊñáÊµãËØï"
        unicode_trigger = "–°–æ–±—ã—Ç–∏–µ_‰∏≠Êñá_üéØ"

        dto = FsmSnapshotDTO(
            version=1,
            state=FsmState.IDLE,
            reason=unicode_reason,
            context_data={"unicode_key": "–∑–Ω–∞—á–µ–Ω–∏–µ_‰∏≠Êñá"},
            state_metadata={"—Å–æ–±—ã—Ç–∏–µ": unicode_trigger},
        )

        proto = dto_to_proto(dto)
        converted_back = proto_to_dto(proto)

        assert converted_back.reason == unicode_reason
        assert converted_back.context_data["unicode_key"] == "–∑–Ω–∞—á–µ–Ω–∏–µ_‰∏≠Êñá"
        assert converted_back.state_metadata["—Å–æ–±—ã—Ç–∏–µ"] == unicode_trigger

    def test_empty_collections_handling(self):
        """–¢–µ—Å—Ç –æ–±—Ä–∞–±–æ—Ç–∫–∏ –ø—É—Å—Ç—ã—Ö –∫–æ–ª–ª–µ–∫—Ü–∏–π"""
        dto = FsmSnapshotDTO(
            version=1,
            state=FsmState.IDLE,
            history=[],  # –ø—É—Å—Ç–∞—è –∏—Å—Ç–æ—Ä–∏—è
            context_data={},  # –ø—É—Å—Ç—ã–µ –¥–∞–Ω–Ω—ã–µ
            state_metadata={},  # –ø—É—Å—Ç—ã–µ –º–µ—Ç–∞–¥–∞–Ω–Ω—ã–µ
        )

        proto = dto_to_proto(dto)
        converted_back = proto_to_dto(proto)

        assert converted_back.history == []
        assert converted_back.context_data == {}
        # state_metadata –º–æ–∂–µ—Ç —Å–æ–¥–µ—Ä–∂–∞—Ç—å DTO-—Å–ø–µ—Ü–∏—Ñ–∏—á–Ω—ã–µ –ø–æ–ª—è

    def test_large_collections_handling(self):
        """–¢–µ—Å—Ç –æ–±—Ä–∞–±–æ—Ç–∫–∏ –±–æ–ª—å—à–∏—Ö –∫–æ–ª–ª–µ–∫—Ü–∏–π"""
        # –ë–æ–ª—å—à–∞—è –∏—Å—Ç–æ—Ä–∏—è
        large_history = [
            create_transition(FsmState.IDLE, FsmState.ACTIVE, f"EVENT_{i}")
            for i in range(1000)
        ]

        # –ú–Ω–æ–≥–æ –º–µ—Ç–∞–¥–∞–Ω–Ω—ã—Ö
        large_context = {f"key_{i}": f"value_{i}" for i in range(500)}

        dto = FsmSnapshotDTO(
            version=1,
            state=FsmState.ACTIVE,
            history=large_history,
            context_data=large_context,
        )

        proto = dto_to_proto(dto)
        converted_back = proto_to_dto(proto)

        assert len(converted_back.history) == 1000
        assert len(converted_back.context_data) == 500
        assert all(h.trigger_event.startswith("EVENT_") for h in converted_back.history)


if __name__ == "__main__":
    # –ë—ã—Å—Ç—Ä—ã–π –∑–∞–ø—É—Å–∫ —Ç–µ—Å—Ç–æ–≤
    pytest.main([__file__, "-v", "--tb=short"])
